File: numass-control/src/main/java/inr/numass/control/NumassControlApplication.java
package inr.numass.control;

import ch.qos.logback.classic.Level;
import hep.dataforge.context.Context;
import hep.dataforge.control.connections.DeviceConnection;
import hep.dataforge.control.connections.Roles;
import hep.dataforge.control.devices.Device;
import hep.dataforge.control.devices.DeviceFactory;
import hep.dataforge.exceptions.ControlException;
import hep.dataforge.meta.Meta;
import javafx.application.Application;
import javafx.application.Platform;
import javafx.scene.Scene;
import javafx.stage.Stage;
import org.slf4j.LoggerFactory;

import java.util.Locale;

/**
 * Created by darksnake on 14-May-17.
 */
public abstract class NumassControlApplication<D extends Device> extends Application {
    private D device;

    @Override
    public void start(Stage primaryStage) throws Exception {
        Locale.setDefault(Locale.US);// чтобы отделение десятичных знаков было точкой
        ch.qos.logback.classic.Logger rootLogger = (ch.qos.logback.classic.Logger) LoggerFactory.getLogger(ch.qos.logback.classic.Logger.ROOT_LOGGER_NAME);
        rootLogger.setLevel(Level.INFO);

        DeviceViewConnection<D> controller = buildView();

        Scene scene = new Scene(controller.getPane());

        primaryStage.setScene(scene);
        primaryStage.show();

        device = setupDevice(controller);
        setupStage(primaryStage, device);
    }

    /**
     * Build a view connection
     *
     * @return
     */
    protected abstract DeviceViewConnection<D> buildView();

    /**
     * Get a device factory for given device
     *
     * @return
     */
    protected abstract DeviceFactory<D> getDeviceFactory();

    protected abstract void setupStage(Stage stage, D device);

    protected abstract boolean acceptDevice(Meta meta);

    private D setupDevice(DeviceConnection<D> controller) {
        Meta config = NumassControlUtils.getConfig(this)
                .orElseGet(() -> NumassControlUtils.readResourceMeta("/config/devices.xml"));

        Context ctx = NumassControlUtils.setupContext(config);
        Meta deviceConfig = NumassControlUtils.findDeviceMeta(config, this::acceptDevice)
                .orElseThrow(() -> new RuntimeException("Device configuration not found"));


        try {
            D d = getDeviceFactory().build(ctx, deviceConfig);
            d.init();
            NumassControlUtils.connectStorage(d, config);
            Platform.runLater(() -> {
                d.connect(controller, Roles.VIEW_ROLE, Roles.DEVICE_LISTENER_ROLE);
            });
            return d;
        } catch (ControlException e) {
            throw new RuntimeException("Failed to build device", e);
        }
    }

    @Override
    public void stop() throws Exception {
        super.stop();
        if (device != null) {
            device.shutdown();
            device.getContext().close();
        }
    }


}

File: numass-control/src/main/kotlin/inr/numass/control/NumassControlApplication.kt
package inr.numass.control

import ch.qos.logback.classic.Level
import hep.dataforge.control.devices.Device
import hep.dataforge.control.devices.DeviceFactory
import hep.dataforge.exceptions.ControlException
import hep.dataforge.meta.Meta
import hep.dataforge.optional
import javafx.scene.Scene
import javafx.stage.Stage
import org.slf4j.LoggerFactory
import tornadofx.*
import java.util.*

/**
 * Created by darksnake on 14-May-17.
 */
abstract class NumassControlApplication<in D : Device> : App() {
    private var device: D? = null

    override fun start(stage: Stage) {
        Locale.setDefault(Locale.US)// чтобы отделение десятичных знаков было точкой
        val rootLogger = LoggerFactory.getLogger(ch.qos.logback.classic.Logger.ROOT_LOGGER_NAME) as ch.qos.logback.classic.Logger
        rootLogger.level = Level.INFO

        device = setupDevice().also {
            val controller = it.getDisplay()
            val scene = Scene(controller.view?.root ?: controller.getBoardView())
            stage.scene = scene

            stage.show()
            setupStage(stage, it)
            setDFStageIcon(stage)
        }
    }

    /**
     * Get a device factory for given device

     * @return
     */
    protected abstract val deviceFactory: DeviceFactory

    protected abstract fun setupStage(stage: Stage, device: D)


    abstract fun getDeviceMeta(config: Meta): Meta

    private fun setupDevice(): D {
        val config = getConfig(this).optional.orElseGet { readResourceMeta("config/devices.xml") }

        val ctx = setupContext(config)
        val deviceConfig = getDeviceMeta(config)


        try {
            @Suppress("UNCHECKED_CAST")
            val d = deviceFactory.build(ctx, deviceConfig) as D
            d.init()
            d.connectStorage(config)

            return d
        } catch (e: ControlException) {
            throw RuntimeException("Failed to build device", e)
        }

    }

    override fun stop() {
        try {
            device?.shutdown()
        } catch (ex: Exception) {
            LoggerFactory.getLogger(javaClass).error("Failed to properly shutdown application", ex);
            device?.context?.close()
        } finally {
            super.stop()
        }
    }


}
